src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,244 @@
+        if (!s)
+                return -ENOMEM;
+
+        *ret_solutions = TAKE_PTR(s);
+
+        return 0;
+}
+
+static int query_named_solution(sd_device *device, const char *name, const char **model, const char **qos) {
+        _cleanup_free_ char *upper_name = NULL, *qos_key = NULL, *model_key = NULL;
+        _cleanup_strv_free_ char **solutions = NULL;
+        int r;
+
+        /* If NULL is passed we query the default solution, which is the first one listed
+         * in the SOLUTIONS key.
+         */
+        if (!name) {
+                r = get_known_solutions(device, &solutions);
+
+                if (r < 0)
+                        return log_error_errno(r, "Failed to query solutions from device: %m");
+
+                if (!solutions[0]) {
+                        log_error("IOCOST_SOLUTIONS exists in hwdb but is empty.");
+                        return -EINVAL;
+                }
+
+                name = solutions[0];
+        }
+
+        upper_name = strdup(name);
+        if (!upper_name)
+                return log_oom();
+
+        ascii_strupper(upper_name);
+
+        qos_key = strjoin("IOCOST_QOS_", upper_name);
+        if (!qos_key)
+                return log_oom();
+
+        model_key = strjoin("IOCOST_MODEL_", upper_name);
+        if (!model_key)
+                return log_oom();
+
+        r = sd_device_get_property_value(device, model_key, model);
+        if (r < 0)
+                return log_error_errno(r, "Model key missing from hwdb for iocost solution '%s': %m", name);
+
+        r = sd_device_get_property_value(device, qos_key, qos);
+        if (r < 0)
+                return log_error_errno(r, "QoS key missing from hwdb for iocost solution '%s': %m", name);
+
+        return 0;
+}
+
+static int apply_solution_for_path(const char *path, const char *name) {
+        _cleanup_(sd_device_unrefp) sd_device *device = NULL;
+        _cleanup_free_ char *model = NULL, *qos = NULL;
+        const char *model_params = NULL, *qos_params = NULL;
+        dev_t devnum;
+        int r;
+
+        r = sd_device_new_from_path(&device, path);
+        if (r < 0)
+                return log_error_errno(r, "Error looking up device: %m");
+
+        r = query_named_solution(device, name, &model_params, &qos_params);
+        if (r < 0)
+                return r;
+
+        r = sd_device_get_devnum(device, &devnum);
+        if (r < 0)
+                return log_error_errno(r, "Error getting devnum for device %s: %m", path);
+
+        if (asprintf(&model, DEVNUM_FORMAT_STR " model=linear ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), model_params) < 0)
+                return log_oom();
+
+        if (asprintf(&qos, DEVNUM_FORMAT_STR " enable=1 ctrl=user %s", DEVNUM_FORMAT_VAL(devnum), qos_params) < 0)
+                return log_oom();
+
+        log_debug("Applying iocost parameters to %s using solution '%s'\n\tio.cost.model: %s\n\tio.cost.qos: %s\n", path, name ? name : "default", model, qos);
```

Comment:
please also add a C newline here. i.e.:

```c
log_debug("Applying iocost parameters to %s using solution '%s'\n"
                "\tio.cost.model: %s\n"
                "\tio.cost.qos: %s\n", path, name ?: "default", model, qos);

Reply:
and please use gcc's extension that allows dropping the middle param to the ternary op, if it's the same as the left param. i.e. `name ?: "default"`